db/db_impl/db_impl_write.cc

```diff
@@ -60,6 +66,145 @@ Status DBImpl::WriteWithCallback(const WriteOptions& write_options,
+        write_thread_.UpdateLastSequence(versions_->LastSequence()) + 1;
+    size_t total_count = 0;
+    size_t total_byte_size = 0;
+    size_t valid_batches = 0;
+    auto stats = default_cf_internal_stats_;
+    if (writer.status.ok()) {
+      SequenceNumber next_sequence = current_sequence;
+      for (auto w : wal_write_group) {
+        if (w->CheckCallback(this)) {
+          if (w->ShouldWriteToMemtable()) {
+            w->sequence = next_sequence;
+            size_t count = WriteBatchInternal::Count(w->batches);
+            next_sequence += count;
+            total_count += count;
+          }
+          valid_batches += 1;
+          total_byte_size = WriteBatchInternal::AppendedByteSize(
+              total_byte_size, WriteBatchInternal::ByteSize(w->batches));
+        }
+      }
+      if (writer.disable_wal) {
+        has_unpersisted_data_.store(true, std::memory_order_relaxed);
+      }
+      write_thread_.UpdateLastSequence(current_sequence + total_count - 1);
+      if (!write_options.disableWAL) {
+        PERF_TIMER_GUARD(write_wal_time);
+        stats->AddDBStats(InternalStats::kIntStatsWriteDoneBySelf, 1);
+        RecordTick(stats_, WRITE_DONE_BY_SELF, 1);
+        if (wal_write_group.size > 1) {
+          stats->AddDBStats(InternalStats::kIntStatsWriteDoneByOther,
+                            wal_write_group.size - 1);
+
+          RecordTick(stats_, WRITE_DONE_BY_OTHER, wal_write_group.size - 1);
+        }
+        writer.status =
+            WriteToWAL(wal_write_group, log_writer, log_used, need_log_sync,
+                       need_log_dir_sync, current_sequence);
+      }
+    }
+    if (!writer.CallbackFailed()) {
+      WriteStatusCheck(writer.status);
+    }
+
+    if (need_log_sync) {
+      mutex_.Lock();
+      MarkLogsSynced(logfile_number_, need_log_dir_sync, writer.status);
+      mutex_.Unlock();
+    }
+    write_thread_.ExitAsBatchGroupLeader(wal_write_group, writer.status);
+  }
+  if (writer.state == WriteThread::STATE_MEMTABLE_WRITER_LEADER) {
+    PERF_TIMER_GUARD(write_memtable_time);
+    assert(writer.ShouldWriteToMemtable());
+    WriteThread::WriteGroup memtable_write_group;
+    write_thread_.EnterAsMemTableWriter(&writer, &memtable_write_group);
+    assert(immutable_db_options_.allow_concurrent_memtable_write);
+    auto version_set = versions_->GetColumnFamilySet();
+    memtable_write_group.running.store(0);
+    for (auto it = memtable_write_group.begin();
+         it != memtable_write_group.end(); ++it) {
+      if (!it.writer->ShouldWriteToMemtable()) {
+        continue;
+      }
+      WriteBatchInternal::AsyncInsertInto(
+          it.writer, it.writer->sequence, version_set, &flush_scheduler_,
+          ignore_missing_faimly, this,
+          &write_thread_.write_queue_);
+    }
+    while (memtable_write_group.running.load(std::memory_order_acquire) > 0) {
+      std::function<void()> work;
+      if (write_thread_.write_queue_.PopFront(work)) {
+        work();
```

Comment:
here the leader works on the writer queue. how does followers help with the work?

Reply:
Followers would do this work in `AwaitState`.